File: backend/internal/model/page.go
package model

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type Page struct {
	ID        uuid.UUID       `json:"id"`
	ParentID  *uuid.UUID      `json:"parent_id"`
	Title     string          `json:"title"`
	Content   json.RawMessage `json:"content"`
	Position  int             `json:"position"`
	Icon      *string         `json:"icon"`
	Tags      []Tag           `json:"tags"`
	CreatedAt time.Time       `json:"created_at"`
	UpdatedAt time.Time       `json:"updated_at"`
	DeletedAt *time.Time      `json:"deleted_at,omitempty"`
}

type PageTreeNode struct {
	ID        uuid.UUID      `json:"id"`
	ParentID  *uuid.UUID     `json:"parent_id"`
	Title     string         `json:"title"`
	Icon      *string        `json:"icon"`
	Position  int            `json:"position"`
	Tags      []Tag          `json:"tags"`
	Children  []PageTreeNode `json:"children"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt *time.Time     `json:"deleted_at,omitempty"`
}

type TrashedPage struct {
	ID        uuid.UUID  `json:"id"`
	Title     string     `json:"title"`
	Icon      *string    `json:"icon"`
	DeletedAt *time.Time `json:"deleted_at"`
}

type CreatePageRequest struct {
	ParentID *uuid.UUID `json:"parent_id"`
	Title    string     `json:"title"`
}

type UpdatePageRequest struct {
	Title   *string          `json:"title,omitempty"`
	Content *json.RawMessage `json:"content,omitempty"`
	Icon    *string          `json:"icon,omitempty"`
	TagIDs  []uuid.UUID      `json:"tag_ids,omitempty"`
}

type MovePageRequest struct {
	ParentID *uuid.UUID `json:"parent_id"`
	Position int        `json:"position"`
}

type DuplicatePageRequest struct {
	Deep bool `json:"deep"`
}

type SearchResult struct {
	ID    uuid.UUID `json:"id"`
	Title string    `json:"title"`
	Icon  *string   `json:"icon"`
}

File: frontend/src/types/page.ts
import type {Tag} from './tag';

export type {Tag};

export interface Page {
  id: string;
  parent_id: string | null;
  title: string;
  content: JSONContent | null;
  icon: string | null;
  icon_color: string | null;
  position: number;
  tags: Tag[];
  created_at: string;
  updated_at: string;
  deleted_at?: string;
}

export interface PageTreeNode {
  id: string;
  parent_id: string | null;
  title: string;
  icon: string | null;
  icon_color: string | null;
  position: number;
  tags: Tag[];
  children: PageTreeNode[];
  created_at: string;
  updated_at: string;
}

export interface TrashedPage {
  id: string;
  title: string;
  icon: string | null;
  icon_color: string | null;
  deleted_at: string;
}

export interface JSONContent {
  type?: string;
  attrs?: Record<string, unknown>;
  content?: JSONContent[];
  marks?: { type: string; attrs?: Record<string, unknown> }[];
  text?: string;
}

export interface SearchResult {
  id: string;
  title: string;
  icon: string | null;
  icon_color: string | null;
}

export interface CreatePageRequest {
  parent_id?: string | null;
  title?: string;
}

export interface UpdatePageRequest {
  title?: string;
  content?: JSONContent;
  icon?: string;
  icon_color?: string;
  tag_ids?: string[];
}

export interface MovePageRequest {
  parent_id: string | null;
  position: number;
}
